Improving axes grouping/splitting in static viz.

adam-james-v opened the pull request:
fixes: #20559

### Before
We see that static viz renders with the axis split (we don't want to see the axis labels on the right):

<img width="400" alt="image" src="https://user-images.githubusercontent.com/21064735/223522526-8eaeb8b9-628b-44bf-b641-599eea40a49b.png">

### After
The axis labels are correctly being rendered only on the left:

<img width="400" alt="image" src="https://user-images.githubusercontent.com/21064735/223522811-50c73f31-ff93-4298-b011-ded240f006bb.png">

Additionally:
- if a user explicitly sets any of the x-axis breakouts to display on the left or right, the setting is respected.
- axis grouping/splitting implementation works for both single-x-axis with multiple series and multi-x-axis.

### What's the Difference between Single-X-Axis and Multi-X-Axis?

This confused me at first, so I'll try to explain things. There are 2 ways in which a chart will have multiple lines (or areas or bars) rendered:

#### 1. A multi-x-axis situation.
This arises when you create a question with 2 (currently only 2) x-axes, with one of them having several breakouts. 

For example:
<img width="350" alt="image" src="https://user-images.githubusercontent.com/21064735/223529985-19af9416-fd6c-479b-8576-1a1d9bed893d.png">

Which, in the data section as you set up your visualization, will look like this:
<img width="350" alt="image" src="https://user-images.githubusercontent.com/21064735/223532114-d5f3c7e3-a446-4f08-8dde-4c6af180f680.png">

And, the shape of the rows that the query returns looks like this:

```clojure
;; X-axis      X-axis                Y-axis
[["Doohickey" "2016-01-01T00:00:00Z" 9031.557803943724]
 ["Doohickey" "2017-01-01T00:00:00Z" 43069.138089392625]
 ["Doohickey" "2018-01-01T00:00:00Z" 98515.46952597176]
 ["Doohickey" "2019-01-01T00:00:00Z" 110322.19082203327]
 ["Doohickey" "2020-01-01T00:00:00Z" 36332.63737703899]
 ["Gadget" "2016-01-01T00:00:00Z" 10672.625619920142]
 ["Gadget" "2017-01-01T00:00:00Z" 54960.620367033574]
 ["Gadget" "2018-01-01T00:00:00Z" 133811.6893419144]
 ["Gadget" "2019-01-01T00:00:00Z" 160502.6901169807]
 ["Gadget" "2020-01-01T00:00:00Z" 46673.62035739867]
 ["Gizmo" "2016-01-01T00:00:00Z" 9929.320704594422]
 ["Gizmo" "2017-01-01T00:00:00Z" 48130.70799398427]
 ["Gizmo" "2018-01-01T00:00:00Z" 136779.80350954592]
 ["Gizmo" "2019-01-01T00:00:00Z" 145044.8403051793]
 ["Gizmo" "2020-01-01T00:00:00Z" 43177.27364199722]
 ["Widget" "2016-01-01T00:00:00Z" 12523.365117960406]
 ["Widget" "2017-01-01T00:00:00Z" 59095.5552293035]
 ["Widget" "2018-01-01T00:00:00Z" 140938.0718003145]
 ["Widget" "2019-01-01T00:00:00Z" 161197.3266454905]
 ["Widget" "2020-01-01T00:00:00Z" 49913.1786806334]]
```

Where each row is one of the combinations of the X-Axis values, followed by the Y-Axis value for that combination. So, that first row is: _The sum of total orders of the product 'Doohickey' for the year 2016_.

It's worth recognizing that this is distinct from the next situation:

#### 2. Single-x-axis, with Multiple Series situation:
This was a bit easier for me to follow. This is when there is only 1 x-axis but any number of Y-Axes. For example (ignoring whether or not the question setup is sensible):

<img width="350" alt="image" src="https://user-images.githubusercontent.com/21064735/223538116-5bef75a9-7fda-4d83-81ae-d60b55e074c5.png">

Those 3 'summarize' entries can be added to the Y-Axis section when setting up your visualization:
<img width="350" alt="image" src="https://user-images.githubusercontent.com/21064735/223538861-f9050280-6e59-41b2-9dfb-8c9dcdc24ad7.png">

And the query returns the rows in the following shape:

```clojure
;; X-Axis value          Y-axis values
[["2022-04-03T00:00:00Z" 55 18 55]
 ["2022-04-04T00:00:00Z" 476 89 531]
 ["2022-04-05T00:00:00Z" 459 103 990]
 ["2022-04-06T00:00:00Z" 474 93 1464]
 ["2022-04-07T00:00:00Z" 455 99 1919]
 ["2022-04-08T00:00:00Z" 785 94 2704]
 ["2022-04-09T00:00:00Z" 87 26 2791]
 ["2022-04-10T00:00:00Z" 152 22 2943]
 ["2022-04-11T00:00:00Z" 548 92 3491]
 ["2022-04-12T00:00:00Z" 472 87 3963]
 ["2022-04-13T00:00:00Z" 641 116 4604]
 ["2022-04-14T00:00:00Z" 369 80 4973]
 ["2022-04-15T00:00:00Z" 265 56 5238]
 ["2022-04-16T00:00:00Z" 76 17 5314]
 ["2022-04-17T00:00:00Z" 6 4 5320]]
```

adam-james-v committed:
WIP for better axes grouping in static viz.

Double x-axis and single x-axis are different, but both can have situations where grouping might occur, so the
implementation is a tad confusing. I'm trying to build test cases that represent both scenarios reasonably, and build
an implementation that's not too confusing.

adam-james-v committed:
Merge branch 'master' into bugfix-static-viz-multi-series-shows-two-y-axis-20559

adam-james-v committed:
Merge branch 'master' into bugfix-static-viz-multi-series-shows-two-y-axis-20559

adam-james-v committed:
Merge branch 'master' into bugfix-static-viz-multi-series-shows-two-y-axis-20559

adam-james-v committed:
Fix reflection warnings in render test utils

adam-james-v committed:
Merge branch 'master' into bugfix-static-viz-multi-series-shows-two-y-axis-20559

adam-james-v committed:
Merge branch 'master' into bugfix-static-viz-multi-series-shows-two-y-axis-20559

adam-james-v committed:
Overlap calcs don't fail with BigDecimal inputs now.

adam-james-v committed:
Continuing to clean up overlaps calcs and impl for single/double in progress

adam-james-v committed:
WIP version that works for single-x multi-series and double-x multi

There's a lot to think about here...

But basically, handle rows of shape:

double-x:

[[x1 x2] [y]]

and single-x:

[[x] [y1 y2 y3]]

adam-james-v committed:
Single and Multi X axis now use same axes grouping strategy

adam-james-v committed:
Cover some overlap scoring cases.

This may need improving in future still, but the overlap calculation is at least used for both single-x and multi-x
axis scenarios now, which is a definitie improvement.

adam-james-v committed:
Merge branch 'master' into bugfix-static-viz-multi-series-shows-two-y-axis-20559

adam-james-v committed:
Missed these unused bindings in test let body

adam-james-v committed:
Merge branch 'master' into bugfix-static-viz-multi-series-shows-two-y-axis-20559

adam-james-v committed:
overlap calc won't fail if we pass in `[nil nil]`

noahmoss commented on src/metabase/pulse/render/body.clj:
```diff
@@ -549,23 +549,53 @@
+      (and (= min-a max-a) (<= min-b min-a max-b)) 1
+      (and (= min-b max-b) (<= min-a min-b max-a)) 1
+
+      ;; if we have a range that is a single point, is outside the other range, but
+      ;; is near it, we will calculate a 'nearness', which we will consider an overlap.
```
Adding the concept of `nearness` sort of contradicts the name of the function being `overlap` which could be confusing. It's also a pretty long function so might be worth refactoring.

Here's a quick suggestion: split this out into two separate `overlap` and `nearness` functions and have them both called by a more generically-named function like `axis-group-threshold-val`.

noahmoss commented on src/metabase/pulse/render/body.clj:
```diff
@@ -642,41 +672,95 @@
+            ;; If there are 4 groups, there will be 4x4 overlaps calculated.
+            ;; TODO: the reciprocal pairings are technically unneeded calculation. It's optimizable,
+            ;; but the number of total groupings might never be high enough to really worry about it.
+            overlaps           (-> joined-rows-map
+                                   (update-vals (fn [vs]
+                                                  (into {} (map (fn [k]
+                                                                  [k (overlap (get joined-rows-map k) vs)])
+                                                                (keys joined-rows-map))))))
```
Looks like the only use of `overlaps` is to get `relevant-distances`. Can't you just compute `relevant-distances` directly and skip creating the entire 2-d map of overlaps?

adam-james-v replied:
Yea, that's smart. Haha, thanks for pointing it out. I do that with `score-fn` instead now. Simpler!

noahmoss commented on src/metabase/pulse/render/body.clj:
```diff
@@ -642,41 +672,95 @@
+            lefts              (or (:left positions) [(first (get positions nil))])
+            rights             (or (:right positions) [])
+            to-group           (remove (set (concat lefts rights)) (get positions nil))
+            relevant-distances (get overlaps (first lefts))
```
nit: just call this `relevant-overlaps`? (or just `overlaps` if you address my comment above). probably best to use consistent terminology, and `distances` doesn't seem to be used elsewhere

adam-james-v replied:
Good catch. I call things 'score' now, and I removed the creation of the whole map.

noahmoss commented on src/metabase/pulse/render/body.clj:
```diff
@@ -642,41 +672,95 @@
+                               chart-type
+                               (nth default-combo-chart-types idx))
+            selected-rows  (mapv #(vector (ffirst %) (nth (second %) idx)) joined-rows)
+            y-axis-pos     (or (get positions y-col-key) "left")]
```
nit: `get` can take a default
```suggestion
            y-axis-pos     (get positions y-col-key "left")]
```

adam-james-v replied:
I always forget that. Thanks :)

noahmoss commented on test/metabase/pulse/render/body_test.clj:
```diff
@@ -414,28 +415,28 @@
          (render-bar-graph {:cols         default-columns
                             :rows         [[10.0 1] [5.0 10] [2.50 20] [1.25 30]]
-                            :viz-settings {}}))))
+                            :viz-settings {:graph.metrics ["NumPurchased"]}}))))
```
Curious why these all needed to be updated?

adam-james-v replied:
In a real question, if viz. has been set for line, area, or bar, the :viz-settings key will have :graph.metrics and :graph.dimensions values. 

The axis grouping strategy requires those keys to be able to actually re-construct the series. For example, a row `[timestamp 1 2 3 4 5]` needs to know that each of those values corresponds to a metric.

noahmoss reviewed (changes requested):
Nice, this is complex stuff and looks solid overall. 👍 Requesting changes since I think I have a decent number of concrete asks in the comments but hopefully nothing is too onerous.

adam-james-v committed:
Clean up grouping axes implementation addressing review feedback.

- overlap and nearness split into two functions
- add the fn 'axis-group-score' to call overlap/nearness appropriately, and handle condition checks (eg. nil vals)
- removed old group-axes and now unnecessasry default-y-axis-pos fn -> handled by group-axes-at-once fn
- axis group threshold not passed as arg anymore

adam-james-v commented on src/metabase/pulse/render/body.clj:
```diff
@@ -549,23 +549,53 @@
-                              ;; ranges with same min and max won't be considered ranges.
-                              (not= min-a max-a)
-                              (not= min-b max-b))
+  [vals-a vals-b]
```
Good catch! Changed the docstring

adam-james-v commented on src/metabase/pulse/render/body.clj:
```diff
@@ -642,41 +672,95 @@
+                                  (map-indexed (fn [idx k]
+                                                 [k (mapv #(get (second %) idx) joined-rows)]))
+                                  (into {})))
+        ;; map of group-key -> :left :right or nil
```
Yea, really good point. I used the key `:unassigned` as you suggested.

adam-james-v commented on src/metabase/pulse/render/body.clj:
```diff
@@ -642,41 +672,95 @@
+      {:unassigned? true :stacked? true} (into {} (map (fn [k] [k :left]) (keys joined-rows-map)))
+      {:unassigned? false :stacked? true} (into {} (map (fn [k] [k :left]) (keys joined-rows-map)))
+      {:unassigned? true :stacked? false}
+      (let [;; calculate the overlaps between every group.
```
Yea, fair enough. I tried to clean things up and comment a bit more.

adam-james-v commented on src/metabase/pulse/render/body.clj:
```diff
@@ -642,41 +672,95 @@
+        positions          (-> (group-by second starting-positions)
+                               (update-vals #(mapv first %)))
+        check              {:unassigned? (contains? positions nil)
+                            :stacked?    (boolean (:stackable.stack_type viz-settings))} ]
```
I realized I could use a cond instead for one fewer branches. Added comments to try clarify each cond as well.

adam-james-v committed:
Merge branch 'master' into bugfix-static-viz-multi-series-shows-two-y-axis-20559

noahmoss commented on src/metabase/pulse/render/body.clj:
```diff
@@ -549,50 +549,59 @@
-        (conj (repeat "right")
-              "left")))))
+  [vals-a vals-b]
+  (let [[min-a max-a]          (-> vals-a sort ((juxt first last)))
+        [min-b max-b]          (-> vals-b sort ((juxt first last)))
```
nit: alignment

noahmoss commented on src/metabase/pulse/render/body.clj:
```diff
@@ -738,16 +801,16 @@
+                        chart-type)
+        card-name     (:name card)
         ;; NB: There's a hardcoded limit of arity 2 on x-axis, so there's only the 1-axis or 2-axis case
-        series-seqs     [(if (= (count x-cols) 1)
-                           (single-x-axis-combo-series enforced-type joined-rows x-cols y-cols data card-name)
-                           (double-x-axis-combo-series enforced-type joined-rows x-cols y-cols data card-name))]
-        labels          (combo-label-info x-cols y-cols viz-settings)
-        settings        (->ts-viz (first x-cols) (first y-cols) labels viz-settings)]
+        series-seqs   [(if (= (count x-cols) 1)
+                         (single-x-axis-combo-series enforced-type joined-rows x-cols y-cols data card-name)
+                         (double-x-axis-combo-series enforced-type joined-rows x-cols y-cols data card-name))]
+        labels        (combo-label-info x-cols y-cols viz-settings)
+        settings      (->ts-viz (first x-cols) (first y-cols) labels viz-settings)]
```
nit: I think the alignment before was correct here, no?

adam-james-v committed:
Fixed some alignment mistakes

adam-james-v committed:
Merge branch 'bugfix-static-viz-multi-series-shows-two-y-axis-20559' of https://github.com/metabase/metabase into bugfix-static-viz-multi-series-shows-two-y-axis-20559

adam-james-v committed:
Merge branch 'master' into bugfix-static-viz-multi-series-shows-two-y-axis-20559
